Adding snapshot suffix handling to pack of project.json

toddm opened the pull request:
Passing the suffix command-line argument to pack into the project.json parser to replace "-*" in the package version with the correct snapshot value.  This matches dotnet pack behavior.

@alpaix

toddm committed:
Passing the version suffix into the project.json parser to use as a snapshotValue.

alpaix commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -296,7 +296,7 @@ public static bool ProcessProjectJsonFile(PackageBuilder builder, string basePat
         private static void LoadProjectJsonFile(PackageBuilder builder, string path, string basePath, string id, Stream stream, NuGetVersion version, string suffix, Func<string, string> propertyProvider)
         {
-            PackageSpec spec = JsonPackageSpecReader.GetPackageSpec(stream, id, path);
+            PackageSpec spec = JsonPackageSpecReader.GetPackageSpec(stream, id, path, suffix);
```
var

alpaix commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/MSBuildProjectReferenceProvider.cs:
```diff
@@ -191,7 +191,7 @@ private PackageSpec GetPackageSpec(string msbuildProjectPath)
                 if (File.Exists(path))
                 {
-                    result = JsonPackageSpecReader.GetPackageSpec(projectName, path);
+                    result = JsonPackageSpecReader.GetPackageSpec(projectName, path, string.Empty);
```
Here and everywhere below: please use named argument

alpaix replied:
Also I'd prefer having null as a default value. It looks somewhat clearer to me.

alpaix commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -27,21 +27,22 @@ public class JsonPackageSpecReader
         /// <param name="packageSpecPath">file path</param>
-        public static PackageSpec GetPackageSpec(string name, string packageSpecPath)
+        /// <param name="snapshotValue">snapshot suffix value</param>
+        public static PackageSpec GetPackageSpec(string name, string packageSpecPath, string snapshotValue)
```
I think keeping older signature would make this PR shorter.

alpaix commented:
Need tests with non-default snapshot value.

toddm committed:
Simplifying the change by only changing one of the GetPackageSpec() signatures.  Changed the handling of the version in pack so that it carries back to the output package name.  Modified a couple of tests to cover the scenario.

toddm committed:
Adding tests for the snapshot value.

toddm commented:
Thanks, Alex.  I simplified the change by reducing the signature change of GetPackageSpec. In further testing, I found that the version wasn't propagating back to the nupkg version if it was something like "-rc-*".  This is common in the dotnet build so it is worth addressing.  I added a couple new tests to verify the version comes out correctly in both main cases.

alpaix commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecResolver.cs:
```diff
@@ -277,7 +277,7 @@ public PackageSpec PackageSpec
                                     }
 
-                                    project = JsonPackageSpecReader.GetPackageSpec(stream, Name, FullPath);
+                                    project = JsonPackageSpecReader.GetPackageSpec(stream, Name, FullPath, null);
```
`snapshotValue: null`

alpaix commented:
Looks good. One minor thing then :shipit: !

toddm committed:
Adding a name to a parameter for clarity.
